rmw_iceoryx_cpp/src/rmw_response.cpp

```diff
@@ -17,38 +18,173 @@
+      ret = rcutils_system_time_now(&request_header->received_timestamp);
+
+      if (iceoryx_response_header->getSequenceId() ==
+      iceoryx_client_abstraction->sequence_id_ - 1)
+      {
+        user_payload = iceoryx_response_payload;
+        chunk_header = iox::mepoo::ChunkHeader::fromUserPayload(user_payload);
+        ret = RMW_RET_OK;
+      } else {
+        RMW_SET_ERROR_MSG("Got response with outdated sequence number!");
+        *taken = false;
+        ret = RMW_RET_ERROR;
+      }
+    })
+  .or_else(
+    [&](iox::popo::ChunkReceiveResult) {
+      *taken = false;
+      RMW_SET_ERROR_MSG("No chunk in iceoryx_client");
+      ret = RMW_RET_ERROR;
+    });
 
-  RMW_SET_ERROR_MSG("rmw_iceoryx_cpp does not support responses.");
-  return RMW_RET_UNSUPPORTED;
+  if (ret == RMW_RET_ERROR) {
+    return ret;
+  }
+
+  // if fixed size, we fetch the data via memcpy
+  if (iceoryx_client_abstraction->is_fixed_size_) {
+    memcpy(ros_response, user_payload, chunk_header->userPayloadSize());
+  } else {
+    rmw_iceoryx_cpp::deserializeResponse(
+      static_cast<const char *>(user_payload),
+      &iceoryx_client_abstraction->type_supports_,
+      ros_response);
+  }
+  iceoryx_client->releaseResponse(user_payload);
+  *taken = true;
+  std::cout << "Client took response!" << std::endl;
+
+  return ret;
 }
 
-/// @todo Use the new request/response API of iceoryx v2.0 here
 rmw_ret_t
 rmw_send_response(
   const rmw_service_t * service,
   rmw_request_id_t * request_header,
   void * ros_response)
 {
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_ERROR);
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(request_header, RMW_RET_ERROR);
-  RCUTILS_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_ERROR);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(service, RMW_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(request_header, RMW_RET_INVALID_ARGUMENT);
+  RCUTILS_CHECK_ARGUMENT_FOR_NULL(ros_response, RMW_RET_INVALID_ARGUMENT);
+
+  RMW_CHECK_TYPE_IDENTIFIERS_MATCH(
+    rmw_send_response
+    : service, service->implementation_identifier,
+    rmw_get_implementation_identifier(), return RMW_RET_ERROR);
+
+  auto iceoryx_server_abstraction = static_cast<IceoryxServer *>(service->data);
+  if (!iceoryx_server_abstraction) {
+    RMW_SET_ERROR_MSG("service data is null");
+    return RMW_RET_ERROR;
+  }
+
+  auto iceoryx_server = iceoryx_server_abstraction->iceoryx_server_;
+  if (!iceoryx_server) {
+    RMW_SET_ERROR_MSG("iceoryx_server is null");
+    return RMW_RET_ERROR;
+  }
+
+  rmw_ret_t ret = RMW_RET_ERROR;
+
+  auto * iceoryx_request_header = iox::popo::RequestHeader::fromPayload(
+    iceoryx_server_abstraction->request_payload_);
+  /// @todo Why is it not possible to set the sequence id? Is this automatically done? If so,
+  /// we need to compare the user-provided sequence id with the one from the
+  /// 'iceoryx_request_header'
+  // iceoryx_request_header->setSequenceId(request_header->sequence_number);
```

Comment:
@elBoberido Am I right thinking this can be removed and the sequence number is automatically checked under the hood? I would then add an additional check for the sequence id stored in `iceoryx_request_header` vs. user-provided `request_header`.

Reply:
yup